lib/generator.ts

```diff
@@ -542,22 +570,83 @@ export class Generator extends Cli {
     return source
   }
 
+  private sourceLines(source: string) {
+    source = source.replace(/\r\n/g, '\n')
+    source = source.replace(/\n\r/g, '\n')
+    source = source.replace(/\r/g, '\n')
+
+    const lines = source.split('\n')
+    return lines
+  }
+
+  private findDependencies(
+    typings: IDeclarationMap,
+    moduleName: string,
+    shakenTypings: Set<string>,
+    filter: 'exportOnly' | 'allImports',
+  ): void {
+    shakenTypings.add(moduleName)
+    const current = typings[moduleName]
+    const lines = this.sourceLines(current)
+    const refs =
+      filter === 'allImports'
+        ? this.findAllImportedRefs(lines)
+        : this.findExportedRefsOnly(lines)
+    verbose(`"${moduleName}" references ${JSON.stringify(refs)}`)
+
+    for (const ref of refs) {
+      this.findDependencies(typings, ref, shakenTypings, filter)
+    }
+  }
+
+  private findExportedRefsOnly(lines: string[]) {
+    const exports = /export .*? from ['"]([^'"]+)['"]/
+    const refs = lines.map(l => l.match(exports))
+    return refs.filter(m => m !== null).map(m => m[1])
+  }
+
+  private findAllImportedRefs(lines: string[]) {
+    const staticImports = /(from ['"])([^'"]+)(['"])/
+    const inlineImport = /(import\(['"])([^'"]+)(['"]\))/
+    const refs = [
+      ...lines.map(l => l.match(staticImports)),
+      ...lines.map(l => l.match(inlineImport)),
+    ]
+    return refs.filter(m => m !== null).map(m => m[1])
+  }
+
   /**
    * Combines typings into a single declaration source
    */
   private async combineTypings() {
-    const typings = this.loadTypings()
+    let typings = this.loadTypings()
     await this.clearTempDir()
 
     this.moduleNames = Object.keys(typings)
 
     verbose('Combining typings into single file...')
 
-    const sourceParts: string[] = []
-
     Object.entries(typings).forEach(([moduleName, fileSource]) => {
       fileSource = fileSource.replace(/declare /g, '')
       fileSource = this.resolveImportSources(fileSource, moduleName)
+      typings[moduleName] = fileSource
+    })
+
+    const mainFile = this.getMain()
+    const shake = this.getShake()
+
+    if (shake !== 'off') {
+      verbose(`Shaking typeings using the ${shake} strategy.`)
+      const referenced = new Set<string>()
+      this.findDependencies(typings, mainFile, referenced, shake)
+      const shakenTypings = Object.fromEntries(
+        Array.from(referenced).map(ref => [ref, typings[ref]]),
```

Comment:
Would be good to handle case where typiungs[ref] does not exist. It may happen due to a number of reasons - it could be fail on npm-dts side if module name is normalized incorrectly, if some aliases are used in TypeScript, if there is some error in how interfaces are defined or any number of reasons which we might not be aware of yet.